File: merge_sort.js
const fs = require("fs");

// ===== MERGE SORT IMPLEMENTATION ======

function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const mid = Math.floor(arr.length / 2);
    const leftHalf = arr.slice(0, mid);
    const rightHalf = arr.slice(mid);

    const sortedLeft = mergeSort(leftHalf);
    const sortedRight = mergeSort(rightHalf);

    return merge(sortedLeft, sortedRight);
}

function merge(left, right) {
    const result = [];
    let i = 0;
    let j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    while (i < left.length) {
        result.push(left[i]);
        i++;
    }

    while (j < right.length) {
        result.push(right[j]);
        j++;
    }

    return result;
}

function getValuesFromDataset(datasetName) {
    const data = fs.readFileSync("datasets/input-" + datasetName + ".txt", "utf8");
    const lines = data.split("\n");

    const numbers = [];
    for (let line of lines) {
        line = line.trim();
        if (line.length > 0) {
            numbers.push(parseInt(line, 10));
        }
    }

    return numbers;
}

function main() {
    if (process.argv.length !== 3) {
        console.log("Usage: node merge_sort.js <dataset_size>");
        process.exit(1);
    }

    const datasetName = process.argv[2];
    const values = getValuesFromDataset(datasetName);
    console.log(mergeSort(values));
}

main();

File: merge_sort.py
import sys


# ===== MERGE SORT IMPLEMENTATION ======

def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]

    sorted_left = merge_sort(left_half)
    sorted_right = merge_sort(right_half)

    return merge(sorted_left, sorted_right)


def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])

    return result




def get_values_from_dataset(dataset_size):
    numbers = []
    with open(f"datasets/input-{dataset_size}.txt", "r") as file:
        for line in file:
            line = line.strip()
            if line:
                numbers.append(int(line))
    return numbers

def main():
    if len(sys.argv) != 2:
        print("Usage: python marge_sort.py <dataset>")
        sys.exit(1)

    dataset_size = sys.argv[1]
    values = get_values_from_dataset(dataset_size)
    print(merge_sort(values))

if __name__ == "__main__":
    main()
